receiver/statsdreceiver/protocol/statsd_parser.go

```diff
@@ -81,29 +144,51 @@ func buildMetric(metricName string, point *metricspb.Point) *metricspb.Metric {
 	}
 }
 
-func buildPoint(metricType, metricValue string) (*metricspb.Point, error) {
+func buildPoint(parsedMetric *statsDMetric) (*metricspb.Point, error) {
 	point := &metricspb.Point{
 		Timestamp: &timestamppb.Timestamp{
-			Seconds: time.Now().Unix(),
+			Seconds: timeNowFunc(),
 		},
 	}
 
-	switch metricType {
+	switch parsedMetric.statsdMetricType {
 	case "c":
-		// TODO: support both Int64 and Double values.
-		i, err := strconv.ParseInt(metricValue, 10, 64)
+		i, err := strconv.ParseInt(parsedMetric.value, 10, 64)
 		if err != nil {
-			f, err := strconv.ParseFloat(metricValue, 64)
+			f, err := strconv.ParseFloat(parsedMetric.value, 64)
 			if err != nil {
-				return nil, fmt.Errorf("parse metric value string: %s", metricValue)
+				return nil, fmt.Errorf("parse metric value string: %s", parsedMetric.value)
```

Comment:
Looks like the logic that identifies the type of value is used by the second case as well. Could you abstract this out to a separate method and use it to appropriately set the values for a point?

Reply:
Do you have any suggestions for how to cleanly do this? The `metricspb.Point.Value` is of type `isPoint_Value` which is unexported, so I can't simply reuse that. I also need to know in the context of a "c" or "g" type whether the value is a float or int to set the `metricspb.MetricDescriptor_Type`.

Reply:
```
func buildPoint(parsedMetric *statsDMetric) (*metricspb.Point, error) {
	now := &timestamppb.Timestamp{
		Seconds: timeNowFunc(),
	}

	switch parsedMetric.statsdMetricType {
	case "c":
		return buildGaugeOrCounterPoint(parsedMetric, now, func(parsedMetric *statsDMetric, isDouble bool) {
			if isDouble {
				parsedMetric.metricType = metricspb.MetricDescriptor_CUMULATIVE_DOUBLE
				return
			}
			parsedMetric.metricType = metricspb.MetricDescriptor_CUMULATIVE_INT64
		})
	case "g":
		return buildGaugeOrCounterPoint(parsedMetric, now, func(parsedMetric *statsDMetric, isDouble bool) {
			if isDouble {
				parsedMetric.metricType = metricspb.MetricDescriptor_GAUGE_DOUBLE
				return
			}
			parsedMetric.metricType = metricspb.MetricDescriptor_GAUGE_INT64
		})
	}

	return nil, fmt.Errorf("unhandled metric type: %s", parsedMetric.statsdMetricType)
}

func buildGaugeOrCounterPoint(parsedMetric *statsDMetric, now *timestamppb.Timestamp,
	metricTypeSetter func(parsedMetric *statsDMetric, isDouble bool)) (*metricspb.Point, error) {
	var point *metricspb.Point
	var isDouble bool

	i, err := strconv.ParseInt(parsedMetric.value, 10, 64)
	if err != nil {
		f, err := strconv.ParseFloat(parsedMetric.value, 64)
		if err != nil {
			return nil, fmt.Errorf("parse metric value string: %s", parsedMetric.value)
		}
		point = &metricspb.Point{
			Timestamp: now,
			Value: &metricspb.Point_DoubleValue{
				DoubleValue: f,
			},
		}
		isDouble = true
	} else {
		point = &metricspb.Point{
			Timestamp: now,
			Value: &metricspb.Point_Int64Value{
				Int64Value: i,
			},
		}
	}
	metricTypeSetter(parsedMetric, isDouble)
	return point, nil
}
```

What do you think of doing something like this?

Reply:
Added this, thanks!